File: ImperatorToCK3/Source/Imperator/Characters/CharacterAttributes.cpp
#include "CharacterAttributes.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"

Imperator::CharacterAttributes::CharacterAttributes(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void Imperator::CharacterAttributes::registerKeys()
{
	registerKeyword("martial", [this](std::istream& theStream) {
		martial = commonItems::getInt(theStream);
	});
	registerKeyword("finesse", [this](std::istream& theStream) {
		finesse = commonItems::getInt(theStream);
	});
	registerKeyword("charisma", [this](std::istream& theStream) {
		charisma = commonItems::getInt(theStream);
	});
	registerKeyword("zeal", [this](std::istream& theStream) {
		zeal = commonItems::getInt(theStream);
	});
	registerMatcher(commonItems::catchallRegexMatch, commonItems::ignoreItem);
}

File: ImperatorToCK3/Imperator/Characters/CharacterAttributes.cs
using commonItems;

namespace ImperatorToCK3.Imperator.Characters;

public sealed class CharacterAttributes {
	public int Martial { get; set; } = 0;
	public int Finesse { get; set; } = 0;
	public int Charisma { get; set; } = 0;
	public int Zeal { get; set; } = 0;

	public static CharacterAttributes Parse(BufferedReader reader) {
		var attributes = new CharacterAttributes();
		
		var parser = new Parser();
		parser.RegisterKeyword("martial", r => attributes.Martial = r.GetInt());
		parser.RegisterKeyword("finesse", r => attributes.Finesse = r.GetInt());
		parser.RegisterKeyword("charisma", r => attributes.Charisma = r.GetInt());
		parser.RegisterKeyword("zeal", r => attributes.Zeal = r.GetInt());
		parser.IgnoreAndLogUnregisteredItems();
		parser.ParseStream(reader);
		
		return attributes;
	}
}